Dockerfile

```diff
@@ -1,54 +1,54 @@
-FROM debian:9
 
-ENV DEBIAN_FRONTEND=noninteractive \
-    DEBCONF_NONINTERACTIVE_SEEN=true \
-    LC_ALL=C.UTF-8 \
-    LANG=C.UTF-8
+ARG PYTHON_VERSION=3.6.7
 
-COPY . /app
-
-# get dependencies, build, and remove anything we don't need for running jq.
-# valgrind seems to have trouble with pthreads TLS so it's off.
-
-RUN apt-get update && \
-    apt-get install -y \
-        build-essential \
+FROM python:${PYTHON_VERSION}-stretch AS build
+CMD ["bash"]
+ENV LC_ALL=C.UTF-8
+ARG DEBIAN_FRONTEND=noninteractive
+ARG DEBCONF_NONINTERACTIVE_SEEN=true
+RUN apt-get update \
+ && apt-get install -y --no-install-recommends \
         autoconf \
-        libtool \
-        git \
         bison \
-        flex \
-        python3 \
-        python3-pip \
-        wget && \
-    pip3 install pipenv && \
-    (cd /app/docs && pipenv sync) && \
-    (cd /app && \
-        git submodule init && \
-        git submodule update && \
-        autoreconf -i && \
-        ./configure --disable-valgrind --enable-all-static --prefix=/usr/local && \
-        make -j8 && \
-        make check && \
-        make install ) && \
-    (cd /app/modules/oniguruma && \
-        make uninstall ) && \
-    (cd /app && \
-        make distclean ) && \
-    apt-get purge -y \
         build-essential \
-        autoconf \
-        libtool \
-        bison \
-        git \
         flex \
-        python3 \
-        python3-pip && \
-    apt-get autoremove -y && \
-    rm -rf /app/modules/oniguruma/* && \
-    rm -rf /app/modules/oniguruma/.git && \
-    rm -rf /app/modules/oniguruma/.gitignore && \
-    rm -rf /var/lib/apt/lists/* /var/lib/gems
+        git \
+        libtool \
+ && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /app
+
+# Install docs dependencies; we're only copying the Pipfile here,
+# so that this is only run again if dependencies change
+RUN  pip3 install pipenv
+COPY docs/Pipfile docs/Pipfile.lock ./docs/
+RUN  cd docs && pipenv sync
+
+COPY .gitmodules .
+COPY .git        ./.git/
+RUN git submodule init
+RUN git submodule update
+RUN sed -i.bak '/^AM_INIT_AUTOMAKE(\[-Wno-portability 1\.14\])$/s/14/11/' modules/oniguruma/configure.ac
+
+COPY . .
+RUN autoreconf -if
+RUN ./configure --disable-valgrind --with-oniguruma=builtin YACC=/usr/bin/bison --prefix=/build/
```

Comment:
It's better to do `./configure --prefix=/usr/local` ...

Reply:
...which is the default if we're dealing with classic autoconf, so just drop the `--prefix`.

The reason why `--prefix=/build/` could be bad is this path _could_ end up in a compiled binary (e.g. it will look for config files under `/build/etc/` and so on) which is definitely not what we want here.